WordPress/gutenberg
# Rewrite DatePicker, TimePicker and DateTimePicker in TypeScript

## What?

Rewrites the component JavaScript in `packages/components/date-time` in TypeScript.

Part of #35744

## Why?

We're gradually converting all of `@wordpress/components`, and I'm about to do some work on `DateTimePicker`, so I thought I'd be a good citizen and help convert these components.

## How?

I followed steps 1-9 in [this guide](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#refactoring-a-component-to-typescript).

I'll convert the unit tests and storybooks in follow-up PRs as I ran into some issues there and this PR is big enough already.

The only real obstacle I ran into was getting `DayPickerSingleDateController` from `react-dates` to work properly because the types are in `react-dates` but we import `react-dates/lib/components/DayPickerSingleDateController` for performance reasons.

Runtime changes were mostly minimal: I added some null checks here and there and I changed the `update` function in `TimePicker` to be a partial function as that's easier to type.

## Testing Instructions

https://github.com/WordPress/gutenberg/pull/40754 added tests. You can also play around with the _Publish_ popover in the post sidebar to check that there is nothing off.

I'm a total noob at this so go easy on me and let me know if I missed anything! 😅 🙏
